### PR TITLE
Add util functions to fill foveation pattern bitmaps

### PR DESCRIPTION
**Pull request by jorgeag-google**
The first of a series of CL that integrates the commits from:

https://github.com/bjoeris/bigwheels/tree/experimental-foveation

into BW main's branch. 

The series will have 4 CL, this is the first ona that integrates the following commit

* Add util functions to fill foveation pattern bitmaps

### CODE REVIEW

**Comment by apazylbe** [comment]
Thanks Jorge! Any chance we can split this into separate PRs for each commit/meaningful change instead? This much code is a bit daunting to review.

**Comment by jorgeag-google** [comment]
I have split the PR into the original commits.
The last one also included my won addition to integrate.
I am looking into the shader compilation problem on Linux

**Comment by apazylbe** [comment]
Since you already have this in separate commits, would it be possible to split into PRs?
It would be much easier/faster to review, and address review feedback, feedback is supposed to be addressed as separate commits.

**Comment by jorgeag-google** [comment]
I follows Aliya's advice, I update this PR to contain only one of the CL in the original merge.
I create the nexts ones when this gets merged.

**Comment by bjoeris** [comment]
Thanks Jorge for taking care of merging these.

I don't remember exactly what benchmarking you're planning to do, so I just want to make sure these functions are actually needed. I added them specifically to test FDM and VRS *independent* of OpenXR foveation. These let the application use it's own density patterns.

In normal OpenXR foveation, the foveation pattern comes from OpenXR, via the application via `XrSwapchainImageFoveationVulkanFB` in the next chain of `xrEnumerateSwapchainImages`. For this use case, we would want to build a `ShadingRatePattern` around the density image from OpenXR.

**Comment by jorgeag-google** [comment]
Hi Benson,
Yes, the idea is to port all your changes that lead to the `foveation_benchmark` app.
To my undestanding these is a CL that is part of four required ones (correct me if I am worng here)
I want to integrate all four of these:
1. Add util functions to fill foveation pattern bitmaps
1. Add BlitImage (Vulkan Only)
1. Add foveation benchmark
1. Add multisample support for ShadingRatePattern.
Since they are al required to get the app working.

**Review by bjoeris — APPROVED** [review]
lgtm

File: include/ppx/grfx/grfx_shading_rate_util.h [hunk 1]
```diff
@@ -0,0 +1,30 @@
+// Copyright 2022 Google LLC
```
Comment (apazylbe) [inline include/ppx/grfx/grfx_shading_rate_util.h:1]: 2024

File: include/ppx/grfx/grfx_shading_rate_util.h [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+
+#include "ppx/grfx/grfx_shading_rate.h"
+
+namespace ppx {
+namespace grfx {
+
+void FillShadingRateUniformFragmentSize(ShadingRatePatternPtr pattern, uint32_t fragmentWidth, uint32_t fragmentHeight, Bitmap* bitmap);
+void FillShadingRateUniformFragmentDensity(ShadingRatePatternPtr pattern, uint32_t xDensity, uint32_t yDensity, Bitmap* bitmap);
+void FillShadingRateRadial(ShadingRatePatternPtr pattern, float scale, Bitmap* bitmap);
+void FillShadingRateAnisotropic(ShadingRatePatternPtr pattern, float scale, Bitmap* bitmap);
```
Comment (apazylbe) [inline include/ppx/grfx/grfx_shading_rate_util.h:26]: Would you be able to add some function comments for these?

File: src/ppx/grfx/grfx_shading_rate_util.cpp [hunk 3]
```diff
@@ -0,0 +1,72 @@
+// Copyright 2023 Google LLC
```
Comment (apazylbe) [inline src/ppx/grfx/grfx_shading_rate_util.cpp:1]: 2024

**Review by apazylbe — COMMENTED** [review]
thank you! lgtm, just a few comments. Also could you update the PR title and description?

File: include/ppx/grfx/grfx_shading_rate_util.h [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+
+#ifndef ppx_grfx_shading_rate_util_h
+#define ppx_grfx_shading_rate_util_h
+
+#include "ppx/grfx/grfx_shading_rate.h"
+
+namespace ppx {
+namespace grfx {
+// These set of functions create a bitmap.The bitmap can be used to produce
+// a ShadingRatePattern that in tun can be used in a render pass creation to
```
Comment (apazylbe) [inline include/ppx/grfx/grfx_shading_rate_util.h:23]: tun -> turn

**Review by apazylbe — APPROVED** [review]
lgtm, thanks!
By the way, if possible, next time it would be great to address review feedback in separate commits https://github.com/google/bigwheels/blob/main/CONTRIBUTING.md#for-developers-contributing-a-patch

**Pull request merged**